### PR TITLE
m4: modernize recipe + don't add `-rtlib=compiler-rt`

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **m4/1.4.18**

- Modernize recipe
- Also revert https://github.com/conan-io/conan-center-index/commit/7c5762b976f80459026fe79e29c16cf853154412 .
Adding this flag does not belong in m4. It should be put in a profile.
For #6777

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/m4/all/test_package/conanfile.py [hunk 1]
```diff
@@ -2,36 +2,40 @@
...
     def _m4_input_path(self):
         return os.path.join(self.build_folder, "input.m4")
 
     def build_requirements(self):
-        if tools.os_info.is_windows:
-            self.build_requires("msys2/20200517")
+        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            self.build_requires("msys2/cci.latest")
+        if hasattr(self, "settings_build"):
+            self.build_requires(str(self.requires["m4"]))
```
Comment (madebr) [inline recipes/m4/all/test_package/conanfile.py:23]: This is causing bootstrap problems when cross building on c3i.

File: recipes/m4/all/test_package/conanfile.py [hunk 2]
```diff
@@ -2,36 +2,40 @@
...
-        tools.save(self._m4_input_path, M4_CONTENTS)
+        tools.save(self._m4_input_path, textwrap.dedent("""\
+            m4_define(NAME1, `Harry, Jr.')
+            m4_define(NAME2, `Sally')
+            m4_define(MET, `$1 met $2')
+            MET(`NAME1', `NAME2')
+        """))
 
     def test(self):
         m4_bin = tools.get_env("M4")
```
Comment (madebr) [inline recipes/m4/all/test_package/conanfile.py:34]: Guess we can only test `M4` when not cross building, and just use `m4` in other cases.

**Comment by jgsogo** [comment]
Let's say some words about this issue: **Conan v1.x in cross-building scenario** (CCI context).

* We are using `conan install ... --build --profile:host=host --profile:build=build`. This will generate a binary for the _host_ context, as expected. This binary cannot run in the _build_ machine (arch mismatch) so we cannot execute it.
* We are using the same input profiles for testing: `conan test ... --profile:host=host --profile:build=build` and the package being tested is added automatically as `requires` of the testing project in the _host_ context:
   + This guarantees that the binary is available (_host_ context) 
   + This is what you get as input if you run `conan create ... --profile:host=host --profile:build=build` (instead of `conan install --build` + `conan test`.

   The consequence is that the same package for the _build_ context is not available and hence the error (missing binary) if you add it as a build-requires.

---

In CCI we could modify the test call and run `conan test ... --profile:host=host --profile:build=host` (use the same profile for both contexts, the one that was used to build the binary), but it would have two side effects:
 * Now it will find the binary as build-requires, but you cannot run it because it has been built for the _host_ context and you are in the _build_ machine (no big advantage).
 * Input mismatch with users running `conan create` because they will keep using two different profiles, recipe+test_package might fail for them (this can be reported as a bug).

---

So, what can we do in test_package that would work in the general scenario (generic cross-building one)? We cannot trust that a build-requires will run in the _build_ machine (it might be using _host_ settings!), so in general, it is not possible to run a package that we have just build, we can only check that the environment populated is the expected one.

```python

class TestPackage(ConanFile):
    generators = "virtualenv"  # IIRC, this is the one we are looking for.
    
    def build(self):
       if not cross-building:
          # We can actually use our tool:
          self.run("./activate && tool <files>")

    def test(self):
       if not cross-building:
          # Check the tool did the job
       
       # We can always activate the environment and check things
       # - Check PATH contains the package folder
       # - Check 'M4' envvar exists (and value)
       self.run("./activate && check envvars"). 
```

* If we are not cross-building, we can run the binary. We don't really need to add it as a build-requires, we can add the proper _virtualenv_ generator and activate it before running the commands. 
* If we are crossbuilding, we can just activate the environment and check that it now contains the variables that were declared in the `package_info`.

[Adding it as a build-requires](https://docs.conan.io/en/latest/devtools/build_requires.html?highlight=test%20build%20requires#testing-build-requires) would have the same effect, although Conan will activate the environment automatically for you at some places... but still, if it is a _host_ package (running incompatible _build_ platform) it will fail to run.

**Comment by madebr** [comment]
Currently, testing a package as a build requirement when cross building fails because a native package is not available.
I think c3i must build the packages in 2 phases (if needed).
First build the native packages and upload those.
Then build the cross built packages.

**Comment by jgsogo** [comment]
> Currently, testing a package as a build requirement when cross building fails because a native package is not available.
> I think c3i must build the packages in 2 phases (if needed).
> First build the native packages and upload those.
> Then build the cross built packages.

This could be another approach, it is something we can orchestrate in C3i, but not something suitable for a user running `conan create` in their machine. They will get a missing binary and, it would require some knowledge about c3i to realize that they should build the native package first and then run the cross-building create command. 

Also, there is some conceptual error: with that approach, it will be testing the native package, not the cross-compiled one, which can contain different information in the `package_info` method.... and in this job we want to test the cross-compiled package, the native one was already tested in its own build.

The only solution would be to run the tests in a machine matching _host_ settings: spawn an M1 node and run `conan test` there... it is possible for some configurations, but very unlikely for others that might be added in the future (Android, iOS,...)

**Comment by madebr** [comment]
> This could be another approach, it is something we can orchestrate in C3i, but not something suitable for a user running `conan create` in their machine. They will get a missing binary and, it would require some knowledge about c3i to realize that they should build the native package first and then run the cross-building create command.

While building icu in cross build configuration, running `conan create . icu/69.1@ -pr:h android -pr:b native` will automatically build both (build and host).
See https://github.com/conan-io/conan-center-index/pull/6652#discussion_r682172842

> Also, there is some conceptual error: with that approach, it will be testing the native package, not the cross-compiled one, which can contain different information in the `package_info` method.... and in this job we want to test the cross-compiled package, the native one was already tested in its own build.

Indeed, I also brought this up yesterday: https://github.com/conan-io/conan-center-index/pull/6685#discussion_r683840275
This also touches conan's new `test_type = "build_requires"` feature.

> The only solution would be to run the tests in a machine matching _host_ settings: spawn an M1 node and run `conan test` there... it is possible for some configurations, but very unlikely for others that might be added in the future (Android, iOS,...)

**Comment by jgsogo** [comment]
> > This could be another approach, it is something we can orchestrate in C3i, but not something suitable for a user running `conan create` in their machine. They will get a missing binary and, it would require some knowledge about c3i to realize that they should build the native package first and then run the cross-building create command.
> 
> While building icu in cross build configuration, running `conan create . icu/69.1@ -pr:h android -pr:b native` will automatically build both (build and host).
> See [#6652 (comment)](https://github.com/conan-io/conan-center-index/pull/6652#discussion_r682172842)
> 

I know it, I did that change here https://github.com/conan-io/conan-center-index/pull/5796#issue-663723187 and I really think I was using a bug as a feature, but it did the trick and given Conan v1.x it's not going to be fixed (it is wasting resources, because some packages are being built twice).

> we use bug/feature in Conan: as we are creating the packages with --build=icu, in a cross-building scenario Conan will build first the build-requires and then the library itself. We are building twice some packages, but still it works.

It was possible in ICU because the package was a build-require of itself, nevertheless IIRC the `test_package` is only using the _host_ package. It is testing the one it has just built.

IMHO, it is impossible to test a build-requires from a cross-building `conan create`, as the package that has been created cannot run in the build machine... adding the extra feature (dockers, emulators,...) in c3i would require efforts and resources that we need to implement other features that can add more value.

I really think that the best a `test_package` can do in a generic recipe that supports two-profiles and cross-building is to test that the environment populated by the package is the expected one. It can be achieved using an explicit `virtualenv` generator or via `test_type` (I personally prefer the first one as it is more generic for libraries and executables, everything being tested always comes from the _host_ context and it is the [single/unique] bottom vertex of the dependency graph. Anyway, I understand that the second one also tests some built-in features of Conan for that package).

**Comment by madebr** [comment]
@SpaceIm 
Can I get an initial review please?

**Review by SpaceIm — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**